### PR TITLE
Enhance license analysis

### PR DESCRIPTION
**Pull request by harjinder-hari**
This PR is for merging the following changes in license analysis:
- add support for license synonyms
- provide more details in the output e.g. conflicting licenses, unknown licenses
- identify license based outliers

### CODE REVIEW

**Comment by harjinder-hari** [comment]
IMPORTANT: please note that some tests are pending that I can only work on Monday. So, I have tagged this PR as WIP.

PLEASE DO NOT MERGE THIS.

**Comment by harjinder-hari** [comment]
I completed testing on dev cluster and committed the required changes in openshift template. I have also removed the WIP keyword from this PR.

File: src/util/data_store/local_filesystem.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    #     return pd.read_json(os.path.join(self.src_dir, filename))
+    #
+    # def write_pandas_df_into_json_file(self,data,filename):
+    #     data.to_json(os.path.join(self.src_dir, filename))
+
+    @classmethod
+    def byteify(cls, input):
+        if isinstance(input, dict):
+            return {LocalFileSystem.byteify(key): LocalFileSystem.byteify(value)
+                    for key, value in input.iteritems()}
```
Comment (jpopelka) [inline src/util/data_store/local_filesystem.py]: Could we use `items()` to be Python 3 ready ?

Comment (harjinder-hari) [reply src/util/data_store/local_filesystem.py]: Thanks Jiri for the review ! I somehow missed seeing your review comments.

Sure, I will address the review comments and migrate to Python 3.

**Comment by harjinder-hari** [comment]
Hi Jiri (@jpopelka) ,

I have made the migration to Python 3. Can you please review the latest commit ?

Thanks,
Harjinder

File: src/license_analysis.py [hunk 2]
```diff
@@ -141,10 +141,10 @@ def _find_walks_within_type(self, v, lic_type, current_walk):
         current_walk.append(v)
 
         neighbours = v.get_neighbours()
-        neighbours = filter(lambda x: x.get_prop_value('type') == lic_type, neighbours)
+        neighbours = list(filter(lambda x: x.get_prop_value('type') == lic_type, neighbours))
 
         if neighbours is None or len(neighbours) == 0:
-            lic_walk = map(lambda x: x.get_prop_value('license'), current_walk)
+            lic_walk = list(map(lambda x: x.get_prop_value('license'), current_walk))
```
Comment (jpopelka) [inline src/license_analysis.py]: Or
`lic_walk = [x.get_prop_value('license') for x in current_walk]`

File: src/license_analysis.py [hunk 3]
```diff
@@ -315,7 +315,7 @@ def _find_conflict_licenses(self, license_vertices):
         list_groups = list(set(list_groups))
         assert(len(list_groups) > 1)
 
-        list_items = map(lambda x: (x, []), list_groups)
+        list_items = list(map(lambda x: (x, []), list_groups))
```
Comment (jpopelka) [inline src/license_analysis.py]: or
`list_items = [(x, []) for x in list_groups]`

File: src/license_analysis.py [hunk 4]
```diff
@@ -405,10 +405,10 @@ def _find_outlier_licenses(self, license_vertices, stack_license_type):
...
             if self._is_license_stricter(stack_license_type, major_tcc_type):
                 # find all the licenses that fall into same or stricter types
-                items = filter(lambda x:
-                               self._is_license_stricter_or_same(x[1], major_tcc_type),
-                               dict_tcc_type.items())
-                items = filter(lambda x: x[0] != major_tcc_lic, items)
+                items = list(filter(lambda x:
+                                    self._is_license_stricter_or_same(x[1], major_tcc_type),
+                                    dict_tcc_type.items()))
+                items = list(filter(lambda x: x[0] != major_tcc_lic, items))
```
Comment (jpopelka) [inline src/license_analysis.py]: Or
`items = [x for x in dict_tcc_type.items() if self._is_license_stricter_or_same(x[1], major_tcc_type) and x[0] != major_tcc_lic]`

File: tests/test_license_analysis.py [hunk 5]
```diff
@@ -16,33 +25,73 @@ def test_compute_rep_license_successful():
...
+
+def test_compute_rep_license_conflict_2():
+    src_dir = "license_graph"
+    graph_store = LocalFileSystem(src_dir=src_dir)
+    synonyms_dir = "synonyms"
+    synonyms_store = LocalFileSystem(src_dir=synonyms_dir)
+    license_analyzer = LicenseAnalyzer(graph_store, synonyms_store)
+    list_licenses = ['MIT', 'APACHE', 'MPL 1.1', 'lgplv2.1', 'lgplv3+']
+    output = license_analyzer.compute_representative_license(list_licenses)
+    assert output['status'] is 'Conflict'
```
Comment (jpopelka) [inline tests/test_license_analysis.py]: Use `==` for comparing of strings. There are more places where you use `is` for comparing of strings, please fix it. (note: it's perfectly ok to use it to compare against `None`)

File: src/directed_graph.py [hunk 6]
```diff
@@ -19,7 +19,7 @@ def add_neighbor(self, vertex, weight=0):
         self.neighbours[vertex] = weight
 
     def get_neighbours(self):
-        return self.neighbours.keys()
+        return list(self.neighbours.keys())
```
Comment (jpopelka) [inline src/directed_graph.py]: See also `DirectedGraph.get_vertex_ids()`

File: src/license_analysis.py [hunk 7]
```diff
@@ -141,10 +141,10 @@ def _find_walks_within_type(self, v, lic_type, current_walk):
         current_walk.append(v)
 
         neighbours = v.get_neighbours()
-        neighbours = filter(lambda x: x.get_prop_value('type') == lic_type, neighbours)
+        neighbours = list(filter(lambda x: x.get_prop_value('type') == lic_type, neighbours))
```
[the remark this answers is not in the document]
Comment (harjinder-hari) [reply src/license_analysis.py]: Done.

File: src/license_analysis.py (hunk 2 above)
Comment (harjinder-hari) [reply src/license_analysis.py]: Done.

File: src/license_analysis.py (hunk 3 above)
Comment (harjinder-hari) [reply src/license_analysis.py]: Done.

File: src/license_analysis.py (hunk 4 above)
Comment (harjinder-hari) [reply src/license_analysis.py]: Done.

File: tests/test_license_analysis.py (hunk 5 above)
Comment (harjinder-hari) [reply tests/test_license_analysis.py]: Done.

File: src/directed_graph.py (hunk 6 above)
Comment (harjinder-hari) [reply src/directed_graph.py]: Done.

**Comment by harjinder-hari** [comment]
Thanks Jiri (@jpopelka) ! I have used 2to3 tool and your inputs to improve the code further.

Can you please review the latest commit ?

Thanks,
Harjinder

File: src/license_analysis.py [hunk 8]
```diff
@@ -103,13 +103,13 @@ def _print_license_vertex(vertex):
         :param vertex: license vertex to be printed
         :return: None
         """
-        print("Vertex {}: license: {} Type: {}".format(vertex.id,
+        print(("Vertex {}: license: {} Type: {}".format(vertex.id,
```
Comment (jpopelka) [inline src/license_analysis.py]: That's one of the `2to3` false-positives, the print statements are OK as they are.

File: src/license_analysis.py [hunk 9]
```diff
@@ -202,9 +202,9 @@ def _find_type_compatibility_classes(self):
             self._find_walks_within_type(v, lic_type, [])
 
         # deduplicate the member licenses of each type compatibility class
-        for t in self.dict_type_compatibility_classes.keys():
+        for t in list(self.dict_type_compatibility_classes.keys()):
```
Comment (jpopelka) [inline src/license_analysis.py]: Other `2to3` false-positive, i.e. you don't need to convert the 'view' into `list`, your `for` loops are fine as they are.

**Comment by harjinder-hari** [comment]
Hi Jiri ( @jpopelka ),
I have avoided the false positives that you mentioned. Can you please review the latest commit ?
Thanks,
Harjinder

**Comment by miteshvp** [comment]
[test]

**Pull request merged**